### PR TITLE
statistics: remove statistics.Column.Count

### PR DESCRIPTION
**Pull request by xuyifangreeneyes**
### What problem does this PR solve?

Issue Number: ref #42160

Problem Summary:

### What is changed and how it works?

Remove `statistics.Column.Count`. 

Before the PR, in order to maintain `statistics.Column.Count`, we need to read `mysql.stats_top_n` and `mysql.stats_buckets` for each column when init stats, which is time-consuming. On the other hand, the usage of `statistics.Column.Count` is limited. We change `(*DataSource).getColumnNDV` and add `(*Column).StatsCollected` to get rid of `statistics.Column.Count`.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: planner/core/stats.go [hunk 1]
```diff
@@ -163,9 +163,13 @@ func (p *baseLogicalPlan) DeriveStats(childStats []*property.StatsInfo, selfSche
...
-	if ok && hist.Count > 0 {
-		factor := float64(ds.statisticTable.RealtimeCount) / float64(hist.Count)
-		ndv = float64(hist.Histogram.NDV) * factor
+	if ok && hist.IsStatsInitialized() {
+		ndv = float64(hist.Histogram.NDV)
+		// TODO: a better way to get the row count derived from the last analyze.
+		if hist.IsFullLoad() {
+			factor := float64(ds.statisticTable.RealtimeCount) / hist.TotalRowCount()
+			ndv *= factor
+		}
```
Comment (xuyifangreeneyes) [inline planner/core/stats.go:172]: Here the way to calculate column NDV is changed. We assume that the change is ok for most cases.

File: statistics/handle/update.go [hunk 2]
```diff
@@ -998,12 +998,12 @@ var AutoAnalyzeMinCnt int64 = 1000
 // TableAnalyzed checks if the table is analyzed.
 func TableAnalyzed(tbl *statistics.Table) bool {
 	for _, col := range tbl.Columns {
-		if col.Count > 0 {
+		if col.StatsVer != statistics.Version0 {
 			return true
 		}
 	}
 	for _, idx := range tbl.Indices {
-		if idx.Histogram.Len() > 0 {
```
Comment (xuyifangreeneyes) [inline statistics/handle/update.go:1006]: The condition is wrong since `idx.Histogram` may be evicted.

File: statistics/handle/update.go [hunk 3]
```diff
@@ -998,12 +998,12 @@ var AutoAnalyzeMinCnt int64 = 1000
 // TableAnalyzed checks if the table is analyzed.
 func TableAnalyzed(tbl *statistics.Table) bool {
 	for _, col := range tbl.Columns {
-		if col.Count > 0 {
+		if col.StatsVer != statistics.Version0 {
```
Comment (qw4990) [inline statistics/handle/update.go:1001]: Why not use `col.StatsCollected`?

Comment (xuyifangreeneyes) [reply statistics/handle/update.go:1001]: If the column is added and its stats are generated according to its default value, `StatsCollected` returns true but the column is not analyzed. I will add another function called `IsAnalyzed` for better encapsulation.

**Pull request closed without merging**